File: app/cluster_objects/resources.py

import app_config


def get_managed_resources(namespace: str):
    try:
        v1 = app_config.core_v1
        apps_v1 = app_config.apps_v1

        pods = v1.list_namespaced_pod(
            namespace=namespace,
            label_selector="managed-by=k8s-idp"
        )
        pod_info = [{
            "name": pod.metadata.name,
            "ready": all(status.ready for status in pod.status.container_statuses) if pod.status.container_statuses else False
        } for pod in pods.items]

        services = v1.list_namespaced_service(
            namespace=namespace,
            label_selector="managed-by=k8s-idp"
        )
        service_names = [svc.metadata.name for svc in services.items]

        deployments = apps_v1.list_namespaced_deployment(
            namespace=namespace,
            label_selector="managed-by=k8s-idp"
        )
        deployment_info = [{
            "name": dep.metadata.name,
            "ready": dep.status.ready_replicas == dep.status.replicas if dep.status.ready_replicas else False
        } for dep in deployments.items]

        return {
            "pods": pod_info,
            "services": service_names,
            "deployments": deployment_info
        }

    except Exception as e:
        print(f"Error getting resources: {e}")
        return {"error": str(e)}

File: internal/k8s/resources.go
package k8s

import (
	"context"

	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
)

type PodInfo struct {
	Name  string `json:"name"`
	Ready bool   `json:"ready"`
}

type DeploymentInfo struct {
	Name     string `json:"name"`
	Ready    bool   `json:"ready"`
	Replicas int32  `json:"replicas"`
}

type ManagedResources struct {
	Pods        []PodInfo        `json:"pods"`
	Services    []string         `json:"services"`
	Deployments []DeploymentInfo `json:"deployments"`
}

// GetManagedResources lists all k8s-idp managed resources in a namespace
func GetManagedResources(namespace string) (*ManagedResources, error) {
	ctx := context.Background()
	labelSelector := "managed-by=k8s-idp"
	listOpts := metav1.ListOptions{LabelSelector: labelSelector}

	// 1. Get Pods
	pods, err := Client.CoreV1().Pods(namespace).List(ctx, listOpts)
	if err != nil {
		return nil, err
	}
	podInfoList := make([]PodInfo, 0)
	for _, pod := range pods.Items {
		ready := true
		if len(pod.Status.ContainerStatuses) == 0 {
			ready = false
		} else {
			for _, status := range pod.Status.ContainerStatuses {
				if !status.Ready {
					ready = false
					break
				}
			}
		}
		podInfoList = append(podInfoList, PodInfo{Name: pod.Name, Ready: ready})
	}

	// 2. Get Services
	services, err := Client.CoreV1().Services(namespace).List(ctx, listOpts)
	if err != nil {
		return nil, err
	}
	serviceNames := make([]string, 0)
	for _, svc := range services.Items {
		serviceNames = append(serviceNames, svc.Name)
	}

	// 3. Get Deployments
	deployments, err := Client.AppsV1().Deployments(namespace).List(ctx, listOpts)
	if err != nil {
		return nil, err
	}
	deploymentInfoList := make([]DeploymentInfo, 0)
	for _, dep := range deployments.Items {
		ready := false
		var replicas int32 = 1
		if dep.Spec.Replicas != nil {
			replicas = *dep.Spec.Replicas
			if dep.Status.ReadyReplicas == replicas {
				ready = true
			}
		}
		deploymentInfoList = append(deploymentInfoList, DeploymentInfo{Name: dep.Name, Ready: ready, Replicas: replicas})
	}

	return &ManagedResources{
		Pods:        podInfoList,
		Services:    serviceNames,
		Deployments: deploymentInfoList,
	}, nil
}
